token-syncer/src/token_syncer/syncer.clj

```diff
@@ -0,0 +1,187 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns token-syncer.syncer
+  (:require [clojure.set :as set]
+            [clojure.tools.logging :as log]
+            [plumbing.core :as pc]
+            [token-syncer.waiter :as waiter])
+  (:import (org.eclipse.jetty.client HttpClient)))
+
+(defn- successful?
+  "Returns true if the response has a 2XX status code."
+  [{:keys [status]}]
+  (and (integer? status) (<= 200 status 299)))
+
+(defn retrieve-token->url->token-data
+  "Given collections of cluster urls and tokens, retrieve the token description on each cluster.
+   The resulting data structure is a map as follows: token->cluster->token-data where the
+   token-data format is defined by the return value of waiter/load-token."
+  [^HttpClient http-client cluster-urls all-tokens]
+  (pc/map-from-keys
+    (fn [token]
+      (pc/map-from-keys
+        (fn [cluster-url]
+          (waiter/load-token http-client cluster-url token))
```

Comment:
This is a lot of eager loading.  Let's say we have ~5k tokens.  Do we want to make 5k requests, to each cluster?  Could we include the last update time in the token list endpoint to avoid all the extra gets, which won't be required 99% of the time?